sympy/polys/series/python_powerseriesring.py

```diff
@@ -0,0 +1,799 @@
+    max_degree = max(dup_degree(coeffs1), dup_degree(coeffs2))
+
+    if prec1 is None and prec2 is None and (max_degree < ring_prec):
+        series = dup_add(coeffs1, coeffs2, dom)
+        return series, None
+
+    coeffs1, coeffs2, prec = _unify_prec(s1, s2, dom, ring_prec)
+    return dup_add(coeffs1, coeffs2, dom), prec
+
+
+def _useries_sub(
+    s1: USeries[Er], s2: USeries[Er], dom: Domain, ring_prec: int
+) -> USeries[Er]:
+    coeffs1, prec1 = s1
+    coeffs2, prec2 = s2
+    max_degree = max(dup_degree(coeffs1), dup_degree(coeffs2))
+
+    if prec1 is None and prec2 is None and (max_degree < ring_prec):
+        series = dup_sub(coeffs1, coeffs2, dom)
+        return series, None
+
+    coeffs1, coeffs2, prec = _unify_prec(s1, s2, dom, ring_prec)
+    return dup_sub(coeffs1, coeffs2, dom), prec
+
+
+def _useries_mul(
+    s1: USeries[Er], s2: USeries[Er], dom: Domain, ring_prec: int
+) -> USeries[Er]:
+    coeffs1, prec1 = s1
+    coeffs2, prec2 = s2
+
+    if prec1 is None and prec2 is None:
+        d = dup_degree(coeffs1) + dup_degree(coeffs2)
+        if d < ring_prec:
+            return dup_mul(coeffs1, coeffs2, dom), None
+
+    coeffs1, coeffs2, min_prec = _unify_prec(s1, s2, dom, ring_prec)
+    coeffs = dup_series_mul(coeffs1, coeffs2, min_prec, dom)
+    return coeffs, min_prec
+
+
+def _useries_mul_ground(
+    s: USeries[Er], n: Any, dom: Domain, ring_prec: int
+) -> USeries[Er]:
+    coeffs, prec = s
+
+    if n == 0:
+        return [], prec
+    if n == 1:
+        return s
+    if n == -1:
+        return _useries_neg(s, dom, ring_prec)
+
+    series = dup_mul_ground(coeffs, n, dom)
+    return _useries(series, prec, dom, ring_prec)
+
+
+def _useries_div(
+    s1: USeries[Er], s2: USeries[Er], dom: Domain, ring_prec: int
+) -> USeries[Er]:
+    coeffs1, prec1 = s1
+    coeffs2, prec2 = s2
+
+    if not coeffs2:
+        raise ZeroDivisionError("Series division by zero")
+
+    val1 = _useries_valuation(s1, dom)
+    val2 = _useries_valuation(s2, dom)
+
+    if val1 < val2:
+        raise ValueError("quotient would not be a power series")
+
+    if not dom.is_unit(coeffs2[-val2 - 1]):
+        raise ValueError("Trailing coefficient of the divisor must be a unit")
+
+    if val2 == 0:
+        _, _, prec = _unify_prec(s1, s2, dom, ring_prec)
+        s2_inv = _useries_inverse(s2, dom, prec)
+        return _useries_mul(s1, s2_inv, dom, prec)
+    else:
+        # Shift both series to make divisor's valuation zero
+        shifted_coeffs1 = dup_rshift(coeffs1, val2, dom)
+        shifted_coeffs2 = dup_rshift(coeffs2, val2, dom)
+
+        cap = ring_prec - val2
+
+        prec1 = (prec1 if prec1 is not None else ring_prec) - val2
+        prec2 = (prec2 if prec2 is not None else ring_prec) - val2
+
+        shifted_s1 = _useries(shifted_coeffs1, prec1, dom, cap)
+        shifted_s2 = _useries(shifted_coeffs2, prec2, dom, cap)
+
+        q_coeffs = _useries_div(shifted_s1, shifted_s2, dom, cap)[0]
```

Comment:
It is better to avoid having this function call itself recursively like this because it repeats all the checks above. Usually a better way is to have another function so each branch can call that:
```
def _useries_div_inner(...):
   """Assumes both series have valuation 0."""
```